### PR TITLE
Revert "don't need to allocate memory in hasIdentityOpAssign()"

### PR DESCRIPTION
**Pull request by 9rnsr**
Reverts D-Programming-Language/dmd#5446

We must not use stack allocated AST objects, excepting very very few cases.

### CODE REVIEW

**Comment by 9rnsr** [comment]
@WalterBright @andralex @yebblies Please merge this ASAP.

**Comment by ibuclaw** [comment]
Has there been any regressions caused by this?

I agree with this in principle, but if there was a testcase that fails as a result of #5446 then that would give this more weight.

File: src/clone.d [hunk 1]
```diff
@@ -72,11 +72,6 @@ extern (C++) StorageClass mergeFuncAttrs(StorageClass s1, FuncDeclaration f)
 
 /*******************************************
  * Check given aggregate actually has an identity opAssign or not.
- * Params:
- *      ad = struct or class
- *      sc = current scope
- * Returns:
- *      if found, returns FuncDeclaration of opAssign, otherwise null
```
Comment (ibuclaw) [inline src/clone.d]: Don't revert documentation.

File: src/clone.d [hunk 2]
```diff
@@ -371,31 +365,35 @@ extern (C++) FuncDeclaration hasIdentityOpEquals(AggregateDeclaration ad, Scope*
...
-            final switch (i)
-            {
-                case 0:  tthis = ad.type;                 break;
-                case 1:  tthis = ad.type.constOf();       break;
-                case 2:  tthis = ad.type.immutableOf();   break;
-                case 3:  tthis = ad.type.sharedOf();      break;
-                case 4:  tthis = ad.type.sharedConstOf(); break;
-            }
+            if (i == 0)
+                tthis = ad.type;
```
Comment (ibuclaw) [inline src/clone.d]: Could you just revert the bad bits?

File: src/clone.d (hunk 1 above)
Comment (9rnsr) [inline src/clone.d]: I've just used the "revert" button in the PR page, so all changes will be merely reverted.

File: src/clone.d (hunk 2 above)
Comment (9rnsr) [inline src/clone.d]: No, we can open a new PR which contains only refactorings. It's far better.

**Comment by 9rnsr** [comment]
> Has there been any regressions caused by this?

No, it's not found *yet* (I've tried to find it in local quickly, but couldn't).

**Pull request closed without merging**